From 2ac12cfb102550496805ec6d5f9f9d07eded4048 Mon Sep 17 00:00:00 2001
From: Laurent FERT <laurent.fert@intel.com>
Date: Tue, 17 Dec 2013 18:38:51 +0100
Subject: Build: Fix dependency for optimized prebuilt jar

BZ: 158722

Build may fail with an error on missing aapt tool for prebuilt optimized java
libraries.

aapt tool is required by dexpreopt-remove-classes.dex call.
For built libraries, the dependency is not set explicitly on the built module
but comes from a lower level in the dependency chain.
For prebuilt libraries, the dependency must be set explicitly.

Change-Id: Icea237b81425663bc0cbf0784abc80430f898481
Category: aosp improvement
Domain: AOSP-Framework-CoreService
Origin: internal
Upstream-Candidate: no, proprietary
Signed-off-by: Laurent FERT <laurent.fert@intel.com>
---
 core/prebuilt.mk |   31 +++++++++++++++++++++++++++++++
 1 files changed, 31 insertions(+), 0 deletions(-)

diff --git a/core/prebuilt.mk b/core/prebuilt.mk
index 60064ee..31ee0c3 100644
--- a/core/prebuilt.mk
+++ b/core/prebuilt.mk
@@ -204,4 +204,35 @@ $(common_javalib_jar) : $(common_classes_jar) | $(ACP)
 
 # make sure the classes.jar and javalib.jar are built before $(LOCAL_BUILT_MODULE)
 $(built_module) : $(common_javalib_jar)
+
+# Do pre-optimization of the libraries according to build settings.
+# Use the same procedure as java_library.mk for compiled modules.
+ifeq ($(LOCAL_DEX_PREOPT),true)
+dexpreopt_boot_jar_module := $(filter $(LOCAL_MODULE),$(DEXPREOPT_BOOT_JARS_MODULES))
+ifneq ($(dexpreopt_boot_jar_module),)
+# boot jar's rules are defined in dex_preopt.mk
+dexpreopted_boot_jar := $(DEXPREOPT_BOOT_JAR_DIR_FULL_PATH)/$(dexpreopt_boot_jar_module)_nodex.jar
+$(built_module) : $(dexpreopted_boot_jar) | $(ACP)
+	$(call copy-file-to-target)
+
+dexpreopted_boot_odex := $(DEXPREOPT_BOOT_JAR_DIR_FULL_PATH)/$(dexpreopt_boot_jar_module).odex
+built_odex := $(basename $(LOCAL_BUILT_MODULE)).odex
+$(built_odex) : $(dexpreopted_boot_odex) | $(ACP)
+	$(call copy-file-to-target)
+else # dexpreopt_boot_jar_module
+built_odex := $(basename $(LOCAL_BUILT_MODULE)).odex
+$(built_odex): PRIVATE_MODULE := $(LOCAL_MODULE)
+# Make sure the boot jars get dex-preopt-ed first
+$(built_odex) : $(DEXPREOPT_BOOT_ODEXS)
+$(built_odex) : $(common_javalib_jar) | $(DEXPREOPT) $(DEXOPT)
+	@echo "Dexpreopt Jar: $(PRIVATE_MODULE) ($@)"
+	$(hide) rm -f $@
+	@mkdir -p $(dir $@)
+	$(call dexpreopt-one-file,$<,$@)
+
+$(LOCAL_BUILT_MODULE) : $(common_javalib_jar) | $(AAPT) $(ACP)
+	$(call copy-file-to-target)
+	$(call dexpreopt-remove-classes.dex,$@)
+endif # dexpreopt_boot_jar_module
+endif # LOCAL_DEX_PREOPT
 endif # TARGET JAVA_LIBRARIES
-- 
1.7.4.1

